python/spark_sklearn/group_apply.py

```diff
@@ -0,0 +1,158 @@
+    `values` is a `pandas.DataFrame` of columns selected by `cols`, in that order.
+
+    `func` is expected to return a `pandas.DataFrame` of the specified schema `schema`,
+    which should be of type :class:`StructType` (output columns are of this name and order).
+
+    If `spark.conf.get("spark.sql.retainGroupColumns")` is not `u'true'`, then `func` is
+    called without any keys.
+
+    If no `cols` are specified, then all grouped columns will be offered, in the order of the 
+    columns in the original dataframe. In either case, the Pandas columns will be named 
+    according to the DataFrame column names.
+
+    The order of the rows passed in as Pandas rows is not guaranteed to be stable relative to 
+    the original row order.
+
+    .. :note: Users must ensure that the grouped values for every group must fit entirely in
+    memory.
+    .. :note: This method is only available if Pandas is installed.
+
+    :param func: a two argument closure, which may be either a lambda or named function
+    :param schema: the return schema for `func`
+    :param cols: list of column names (string only)
+
+    :raise ValueError: if `*` is in `cols`
+    :raise ValueError: if `cols` contains duplicates
+    :raise ValueError: if `schema` is not a :class:`StructType`
+    :raise ImportError: if `pandas` module is not installed
+    :raise ImportError: if `pandas` version is too old (less than 0.7.1)
+
+    :return: a new :class:`DataFrame` of schema `schema` formed from concatenation of the
+    returned data frames.
+
+    >>> import pandas as pd
+    >>> df = (spark
+    ...     .createDataFrame([Row(course="dotNET", year=2012, earnings=10000),
+    ...                       Row(course="Java",   year=2012, earnings=20000),
+    ...                       Row(course="dotNET", year=2012, earnings=5000),
+    ...                       Row(course="dotNET", year=2013, earnings=48000),
+    ...                       Row(course="Java",   year=2013, earnings=30000)])
+    ...     .select("course", "year", "earnings"))
+    >>> def yearlyMedian(key, vals):
+    ...     all_years = set(vals['year'])
+    ...     # Note that interpolation is performed, so we need to cast back to long.
+    ...     yearly_median = [(key, year, long(vals['earnings'][vals['year'] == year].median()))
+    ...                      for year in all_years]
+    ...     return pd.DataFrame.from_records(yearly_median)
+    >>> gapply(df.groupBy("course"), yearlyMedian, df.schema).orderBy("earnings").show()
+    +------+----+--------+
+    |course|year|earnings|
+    +------+----+--------+
+    |dotNET|2012|    7500|
+    |  Java|2012|   20000|
+    |  Java|2013|   30000|
+    |dotNET|2013|   48000|
+    +------+----+--------+
+    <BLANKLINE>
+    >>> def twoKeyYearlyMedian(course, year, vals):
+    ...     return pd.DataFrame.from_records([(course, year, long(vals["earnings"].median()))])
+    >>> gapply(df.groupBy("course", "year"), twoKeyYearlyMedian, df.schema, "earnings") \
```

Comment:
based on the comment above regarding the keys, then you would not need a special example for multiple keys.